### PR TITLE
unbound: enable more features

### PR DESCRIPTION
**Pull request by poscat0x04**
###### Motivation for this change
Currently unbound in nixpkgs has a lot of not-enabled features. Some of them are quite useful.

###### Things done
This pull request adds options for these features:

- EDNS Client Subnet
- DNSCrypt
- DNSTAP
- TCP Fast Open
- Cache DB module
- Redis backend for the Cache DB module

(also is there a more elegant way of achieving the same goal as [these lines](https://github.com/poscat0x04/nixpkgs/blob/unbound-ecs/pkgs/tools/networking/unbound/default.nix#L123-L125)?)

---

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/tools/networking/unbound/default.nix [hunk 1]
```diff
@@ -97,4 +120,6 @@ stdenv.mkDerivation rec {
     maintainers = with maintainers; [ ehmry fpletz globin ];
     platforms = platforms.unix;
   };
-}
+} // (if !withDNSTAP then {} else {
+  PROTOC_C = "${protobufc}/bin/protoc-c";
+}))
```
Comment (vcunat) [inline pkgs/tools/networking/unbound/default.nix:125]: I'd probably fold this in.  Either with `lib.optionalString` or a `= if foo then bar else null;` (true no-op)

**Comment by vcunat** [comment]
Maybe just enable some of them without introducing an option?  Example candidate: TFO.  

_Though I personally don't have much experience with TFO in DNS.  It seems not common in practice, with either client or server not having it turned on.  Some kernel setting is also needed IIRC._

**Comment by poscat0x04** [comment]
Right, but to my understanding the base unbound is only used for bootstrapping systemd (since there's a dependency cycle between unbound with systemd enabled, gnutls and systemd). unbound-full is what people should use.

**Comment by vcunat** [comment]
On a quick look, the default one for the NixOS service only adds `withSystemd = true`.

**Comment by poscat0x04** [comment]
So... should I leave it as is or should I enable more options? One reason I did not change the base unbound is that changing it will cause a mass rebuild (since systemd depends on it).

**Review by SuperSandro2000 — DISMISSED** [review]

**Review by lovesegfault — DISMISSED** [review]

**Review by lovesegfault — APPROVED** [review]

**Pull request merged**